legate_core_cpp.cmake

```diff
@@ -0,0 +1,410 @@
+endfunction()
+
+if(Legion_USE_Python)
+  _find_package_Python3()
+  if(Python3_FOUND AND Python3_VERSION)
+    set(Legion_Python_Version ${Python3_VERSION})
+  endif()
+endif()
+
+###
+# If we find Legion already configured on the system, it will report whether it
+# was compiled with Python (Legion_USE_PYTHON), CUDA (Legion_USE_CUDA), OpenMP
+# (Legion_USE_OpenMP), and GASNet (Legion_USE_GASNet).
+#
+# We use the same variables as Legion because we want to enable/disable each of
+# these features based on how Legion was configured (it doesn't make sense to
+# build legate.core's Python bindings if Legion's bindings weren't compiled).
+###
+include(cmake/thirdparty/get_legion.cmake)
+
+# If Legion_USE_Python was toggled ON by find_package(Legion), find Python3
+if(Legion_USE_Python AND (NOT Python3_FOUND))
+  _find_package_Python3()
+endif()
+
+if(Legion_USE_GASNet)
+  find_package(MPI REQUIRED)
+endif()
+
+if(Legion_USE_CUDA)
+  # Find the CUDAToolkit
+  rapids_find_package(
+    CUDAToolkit REQUIRED
+    BUILD_EXPORT_SET legate-core-exports
+    INSTALL_EXPORT_SET legate-core-exports
+  )
+
+  # Find NCCL
+  include(cmake/thirdparty/get_nccl.cmake)
+
+  # Enable the CUDA language
+  enable_language(CUDA)
+
+  # Must come after `enable_language(CUDA)`
+  # Use `-isystem <path>` instead of `-isystem=<path>`
+  # because the former works with clangd intellisense
+  set(CMAKE_INCLUDE_SYSTEM_FLAG_CUDA "-isystem ")
+endif()
+
+# Find or install Thrust
+include(cmake/thirdparty/get_thrust.cmake)
+
+##############################################################################
+# - legate.core --------------------------------------------------------------
+
+set(legate_core_SOURCES "")
+set(legate_core_CXX_DEFS "")
+set(legate_core_CUDA_DEFS "")
+set(legate_core_CXX_OPTIONS "")
+set(legate_core_CUDA_OPTIONS "")
+
+include(cmake/Modules/set_cpu_arch_flags.cmake)
+set_cpu_arch_flags(legate_core_CXX_OPTIONS)
+
+if(NOT CMAKE_BUILD_TYPE STREQUAL "Release")
+  list(APPEND legate_core_CXX_DEFS DEBUG_LEGATE)
+  list(APPEND legate_core_CUDA_DEFS DEBUG_LEGATE)
+endif()
+
+if(Legion_USE_CUDA)
+  list(APPEND legate_core_CXX_DEFS LEGATE_USE_CUDA)
+  list(APPEND legate_core_CUDA_DEFS LEGATE_USE_CUDA)
+
+  add_cuda_architecture_defines(legate_core_CUDA_DEFS)
+
+  list(APPEND legate_core_CUDA_OPTIONS -Xfatbin=-compress-all)
+  list(APPEND legate_core_CUDA_OPTIONS --expt-extended-lambda)
+  list(APPEND legate_core_CUDA_OPTIONS --expt-relaxed-constexpr)
+endif()
+
+if(Legion_USE_OpenMP)
+  list(APPEND legate_core_CXX_DEFS LEGATE_USE_OPENMP)
+  list(APPEND legate_core_CUDA_DEFS LEGATE_USE_OPENMP)
+endif()
+
+if(Legion_USE_GASNet)
+  list(APPEND legate_core_CXX_DEFS LEGATE_USE_GASNET)
+  list(APPEND legate_core_CUDA_DEFS LEGATE_USE_GASNET)
+endif()
+
+# Change THRUST_DEVICE_SYSTEM for `.cpp` files
+# TODO: This is what we do in cuNumeric, should we do it here as well?
+if(Legion_USE_OpenMP)
+  list(APPEND legate_core_CXX_OPTIONS -UTHRUST_DEVICE_SYSTEM)
```

Comment:
I ended going in circles trying to understand the behavior of `rapids_cpm_thrust`. I would have thought we could use `rapids_cpm_thrust` to get this with the correct flags already? I couldn't figure out myself how to get thrust to just give you the correct flags. We get a thrust-config.cmake that sets `target_compile_definitions` automatically. If this is how we need to make it work, this approach is fine. Aesthetically would be nice not to have undefine and redefine.

Reply:
Sounds good. Prob better to leave it alone since it works for now and worry about nit clean up after the PR gets merged.